### DISCUSSION TITLE
Porting into_rgb to 0.6

### DISCUSSION THREAD

**Discussion by subpop** [Ogeon/palette#264] [Q&A] [answered]
I am trying to port [some code](https://github.com/pop-os/keyboard-configurator/blob/master/backend/src/color.rs) from 0.5 to 0.6. I am having trouble reasoning my way through the usage of `into_rgb` here, and what `IntoColor` trait I should use instead. I basically have the following snippet:

```rust
let h: f64 = 1.0; 
let hue = RgbHue::from_radians(h);
let hsv = palette::Hsv<palette::encoding::Srgb, f64>::new(hue, *self.s, 1.);
let rgb: palette::LinSrgb<f64> = hsv.into_rgb();
let (r, g, b) = rgb.into_components();
Rgb::from_floats(r, g, b)
```

I understand that `into_rgb` is no longer a method in the `IntoColor` trait, but I'm not understanding what I should replace it with in this conversion. I've tried importing `palette::convert::IntoColor` and using `into_color`, but I get an unsatisfied trait error:

```
the trait bound `palette::rgb::Rgb<palette::encoding::Linear<palette::encoding::Srgb>, f64>: palette::convert::FromColorUnclamped<palette::Hsv<palette::encoding::Srgb, f64>>` is not satisfied

the trait `palette::convert::FromColorUnclamped<palette::Hsv<palette::encoding::Srgb, f64>>` is not implemented for `palette::rgb::Rgb<palette::encoding::Linear<palette::encoding::Srgb>, f64>`

help: the following implementations were found:
        <palette::rgb::Rgb<S, T> as palette::convert::FromColorUnclamped<palette::Alpha<_C, _A>>>
        <palette::rgb::Rgb<S, T> as palette::convert::FromColorUnclamped<palette::Hsl<S, T>>>
        <palette::rgb::Rgb<S, T> as palette::convert::FromColorUnclamped<palette::Hsluv<<<S as palette::rgb::RgbStandard>::Space as palette::rgb::RgbSpace>::WhitePoint, T>>>
        <palette::rgb::Rgb<S, T> as palette::convert::FromColorUnclamped<palette::Hsv<S, T>>>
      and 11 others
note: required because of the requirements on the impl of `palette::convert::IntoColorUnclamped<palette::rgb::Rgb<palette::encoding::Linear<palette::encoding::Srgb>, f64>>` for `palette::Hsv<palette::encoding::Srgb, f64>`
```

Are there restrictions on converting between certain encodings? I'm not intimate with this code, nor how color encoding works in general. I'm just trying to maintain functionality of the code, but with 0.6.

**Comment by Ogeon** [comment, marked as answer]
Hi! You are definitely on the right track, and you are right that the new system doesn't allow converting both color space and encoding in one go. You will have to convert to `Srgb` first, so something like `Srgb::<f64>::from_color(hsv).into_linear()` should do the trick. The reason why this restriction is in place is that removing it makes type inference fail a lot more. I chose to keep it more conservative for now, at the cost of having this blind spot.

**Reply by subpop** [reply]
Thank you! That was exactly the step I was missing. I figured out how to port the reverse conversion too. :tada:

**Reply by Ogeon** [reply]
Nice! 🎉
